### DISCUSSION TITLE
Centroid coordinates for closed ways

### DISCUSSION THREAD

**Discussion by BenOnTrack** [systemed/tilemaker#689] [Q&A] [answered]
I'd like to understand how i can compute with osmium the centroid of a closed way and get back the same result as what's being processed with TileMaker.

Let's take this closed way as an example: https://www.openstreetmap.org/edit?way=23906678#map=19/-36.86075/174.77782
Once my tile is generated with Tilemaker with `LayerAsCentroid`, and I use my favorite library to render my tile, I got 
`[174.77780878543854,-36.86070358695354]`

With Osmium, I've tried 3 methods:

1. compute a linestring with the nodes from the way, and get its centroid (since it looks like what's being done with `calculateCentroid` on Tilemaker) -> result: `[174.77781385463658 -36.860679654659364]`
2. grab the `area` object,  and its outer_ring() to compute a polygon and its centroid -> result: `[174.7778136544017,-36.86070727310713]`
3. using `wkblib` and some of its lib function to also compute the centroid of the polygon -> result: `[174.77781757720027,-36.86076715]`

They all look similar, but none is exactly providing the same result as Tilemaker's. 

I'd need to analyse how much digits really make sense to be kept (maybe 5 are enough?) and how I could truncate tilemaker to the same amounts of digits as I'd need to have exact same values between the tile metadata (tilemaker) and the dataset (osmium).

Happy to hear your thoughts!

**Comment by cldellow** [comment]
tilemaker uses 5 decimal points when writing GeoJSON, so I'd probably use that: https://github.com/systemed/tilemaker/blob/07e878fc5c382a9e1fd9ad255c13f6df533b86f5/include/geojson_writer.h#L147

In v3, the centroid position is calculated using mapbox's polylabel algorithm on the outer ring with the largest area. There are implementations of this in many languages, hopefully you can find one that fits the language you need.

You can also get the centroid in Lua and write it out during tile generation. I wrote a short article about that here: https://www.hikeratlas.com/blog/tilemaker-autosuggest-index

**Reply by BenOnTrack** [reply]
Thanks for your feedback.

I'm now using v3, and I still see the following result for my specific way `[174.77780878543854, -36.86087097675353]` from maplibre-gl with my mbtiles generated with TileMaker. It's 14 decimal points though - I'm using the `queryRenderedFeatures` function to get that result. 

I've downloaded the polylabel function for python (yep, unfortunately not a cpp master) and still got a different result hehe. 

I'm wondering if those data are extrapolated in any way by the libraries I'm using ? My intent is to have:

- the data in mbtiles with its geometry as a Point
- the same data in a .json (from python osmium) stored as a Point as well

and both should return the same geometry, for some fancy functions in the backend :)

**Reply by systemed** [reply, marked as answer]
You can tell tilemaker to use Boost.Geometry's centroid algorithm rather than the Mapbox-originated polylabel one - see CONFIGURATION.md. But ultimately, I suspect the algorithms will always be slightly different between implementations so you won't necessarily get the exact same result.
